Documentation/Patterns.md

```diff
@@ -1,19 +1,130 @@
+    }
+    return menu;
+}
+```
+
+Example from the Kernel:
+
+```cpp
+#include <AK/Try.h>
+
+... snip ...
+
+ErrorOr<Region*> AddressSpace::allocate_region(VirtualRange const& range, StringView name, int prot, AllocationStrategy strategy)
+{
+    VERIFY(range.is_valid());
+    OwnPtr<KString> region_name;
+    if (!name.is_null())
+        region_name = TRY(KString::try_create(name));
+    auto vmobject = TRY(AnonymousVMObject::try_create_with_size(range.size(), strategy));
+    auto region = TRY(Region::try_create_user_accessible(range, move(vmobject), 0, move(region_name), prot_to_region_access_flags(prot), Region::Cacheable::Yes, false));
+    TRY(region->map(page_directory()));
+    return add_region(move(region));
+}
+```
+
+Note: Our `TRY(...)` macro functions similarly to the `?` [operator in rust](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator).
+
+## `MUST(...)` Error Handling
+
+The `MUST(...)` macro is similar to `TRY(...)` except the macro enforces that
+the code run inside the macro must succeed, otherwise we assert.
+
+### Example:
+
+```cpp
+#include <AK/Try.h>
+#include <AK/String.h>
+
+... snip ...
+
+void log_that_can_not_fail(StringView fmtstr, TypeErasedFormatParams& params)
+{
+    StringBuilder builder;
+    MUST(vformat(builder, fmtstr, params));
+    return builder.to_string();
+}
+```
+
+## The `serenity_main(..)` program entry point
+
+Serenity has moved to a pattern where executables do not expose a normal C
+main function. A `serenity_main(..)` is exposed instead. The main reasoning
+is that the `Main::Arguments` struct can provide arguments in a more idiomatic
+way that fits with the serenity API surface area. The ErrorOr<int> likewise
+allows the program to propagate errors seamlessly with the `TRY(...)` macro,
+avoiding a significant amount of clunky C style error handling.
+
+These executables are then linked with the `LibMain` library, which will link in
+the normal C `int main(int, char**)` function which will call into the programs
+`serenity_main(..)` on program startup.
+
+The creation of the pattern was documented in the following video:
+[OS hacking: A better main() for SerenityOS C++ programs](https://www.youtube.com/watch?v=5PciKJW1rUc)
+
+#### Examples: 
+
+A function `main(..)` would normally look something like:
+
+```cpp
+ int main(int argc, char** argv)
+ {
+    return 0;
+ }
+```
+
+Instead, `serenity_main(..)` is defined like this:
+
+```cpp
+#include <LibMain/Main.h>
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    return 0; 
+}
+```
 
 ## Intrusive Lists
 
 [Intrusive lists](https://www.data-structures-in-practice.com/intrusive-linked-lists/) are common in the Kernel and in some specific cases
-are used in the SerenityOS userland. A data structure is said to be
+are used in the SerenityOS user land. A data structure is said to be
```

Comment:
seems like an odd change - by that logic, we'd have to rename `Userland/` to `UserLand/` as well.

Reply:
sure